### PR TITLE
Add GPU ID setting in CLI and FSLeyes

### PR DESCRIPTION
**Pull request by mariehbourget**
## Checklist

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've added relevant tests for my contribution
- [x] I've updated the documentation and/or added correct docstrings
- [x] I've assigned a reviewer
- [x] I've consulted [ADS's internal developer documentation](https://github.com/neuropoly/axondeepseg/wiki/Guidelines-for-contribution) to ensure my contribution is in line with any relevant design decisions

## Description
This PR adds an argument in the CLI and setting in FSLeyes to choose which GPU to use while segmenting.

This feature was already implemented in ivadomed in the `segment_volume` API and is useful when processing large images on our shared GPU clusters.
The default GPU ID `0` is the same as the default in ivadomed.

@Stoyan-I-A, I tried the implementation in the plugin based on the similar implementation of the overlap_value setting. It seems to work fine, let me know if that is ok. We may also need to update the ads_plugin file and/or version after the merge of #700.

## Linked issues
Related to #703.

### CODE REVIEW

**Comment by mathieuboudreau** [comment]
I'll review the CLI behaviour; do we have a someone with a GPU-capable machine that can test the behaviour of the plugin? I can't test that on remote machines and my laptop doesn't have a GPU-compatible card.

**Comment by mariehbourget** [comment]
> do we have a someone with a GPU-capable machine that can test the behaviour of the plugin? I can't test that on remote machines and my laptop doesn't have a GPU-compatible card.

No I don't think we have a GPU-capable machine to test it, I don't either.
To test it, I added a print of the GPU ID in the `axon_segmentation` function to make sure it matches the one from the plugin.

Now maybe adding this to the plugin is a bit overkill as the feature was mainly for us on CLI on the shared GPU-clusters.
If it's not useful and could be confusing for users, we may want to remove it and document it in an issue in case it becomes useful in the future. What do you think?

File: AxonDeepSeg/integrity_test.py
Comment (mariehbourget) [inline AxonDeepSeg/integrity_test.py:35]: Note to the reviewer:
The CI test are not running on GPU, so I did not add a specific test for the GPU ID feature.
However, I added the GPU ID 0 in the integrity test so when we run the tests manually on a GPU cluster, the new feature is taken into account.

Comment (mathieuboudreau) [reply AxonDeepSeg/integrity_test.py:35]: Noted =)

**Comment by mathieuboudreau** [comment]
> > do we have a someone with a GPU-capable machine that can test the behaviour of the plugin? I can't test that on remote machines and my laptop doesn't have a GPU-compatible card.
> 
> No I don't think we have a GPU-capable machine to test it, I don't either. To test it, I added a print of the GPU ID in the `axon_segmentation` function to make sure it matches the one from the plugin.
> 
> Now maybe adding this to the plugin is a bit overkill as the feature was mainly for us on CLI on the shared GPU-clusters. If it's not useful and could be confusing for users, we may want to remove it and document it in an issue in case it becomes useful in the future. What do you think?

I'm fine reviewing the code for this and keeping it. If we get a complaint by a user that it's not working properly, then we'll know who can test the feature for us ;)

**Comment by mariehbourget** [comment]
As discussed in ADS meeting:
* We should add a "count" of the GPU devices available to warn the user if the `--gpu-id `doesn't exist on CLI and to limit the values in the `spinCtrl` in FSLeyes settings.
* There is a bug with GPU allocation, as described by Mathieu in ivadomed issue [1208](https://github.com/ivadomed/ivadomed/issues/1208). The ivadomed team agreed to fix the issue on ivadomed side, so we will need the fix + a new release before we can merge this PR in ADS.

I'll convert the PR to draft as we wait for these to be implemented.

File: AxonDeepSeg/ads_utils.py
Comment (Stoyan-I-A) [inline AxonDeepSeg/ads_utils.py:343]: Not sure if the import will work once FSLeyes has already been loaded. We'll need to test this.

Comment (mariehbourget) [reply AxonDeepSeg/ads_utils.py:343]: I was able to test on rosenberg with X forwarding to my laptop. The display is very slow and by no means usable "in production" but good enough to test that sort of things.
The import seems to work, here is what I did:
1) Open fsleyes and install the plugin
2) Close and reopen fsleyes --> all 8 GPUs were available in settings
3) Close and reopen fsleyes with `CUDA_VISIBLE_DEVICES=0,1,2` --> 3 GPUs were available in settings

I put the import in the function as we usually do with torch in ivadomed. If you foresee any other issues with that for the plugin, we could also put the import at the top of the file. Let me know, thanks!

Comment (Stoyan-I-A) [reply AxonDeepSeg/ads_utils.py:343]: Seems fine to me then 👍

**Comment by mathieuboudreau** [comment]
@mariehbourget are we still waiting for https://github.com/ivadomed/ivadomed/pull/1209 to be merged before changing this out of a draft?

**Comment by mariehbourget** [comment]
> @mariehbourget are we still waiting for [ivadomed/ivadomed#1209](https://github.com/ivadomed/ivadomed/pull/1209) to be merged before changing this out of a draft?

Yes, it would be best to wait for the review to avoid using 2 GPUs when launching the segmentation. It will also require another release from ivadomed first.

**Comment by mathieuboudreau** [comment]
@mariehbourget I saw that the fix was merged in IVADOMED (https://github.com/ivadomed/ivadomed/pull/1209), just not included in a release yet, is that correct?

If there's not an expected release soon, we could move this PR forward by pulling IVADOMED from github in the requirements,

`git+https://github.com/ivadomed/ivadomed`

**Comment by mariehbourget** [comment]
> @mariehbourget I saw that the fix was merged in IVADOMED ([ivadomed/ivadomed#1209](https://github.com/ivadomed/ivadomed/pull/1209)), just not included in a release yet, is that correct?
> 
> If there's not an expected release soon, we could move this PR forward by pulling IVADOMED from github in the requirements,
> 
> `git+https://github.com/ivadomed/ivadomed`

Thanks @mathieuboudreau for the follow-up. Yes, it's fixed in ivadomed and we need a new release but I was sidetrack by more urgent issues. This PR is not very high priority and I'll be on vacation next week so I suggest to postpone to the first week of January where I'll try to release a patch for ivadomed.

**Comment by mathieuboudreau** [comment]
Sounds likes a great plan @mariehbourget ! Happy holidays!

**Comment by mariehbourget** [comment]
Hi @mathieuboudreau and @Stoyan-I-A,
The GPU issue (https://github.com/ivadomed/ivadomed/issues/1208) that was blocking this PR is now corrected with PR #708 and the new ivadomed release (2.9.8).
So this PR is ready for review again.

**Review by mathieuboudreau — APPROVED** [review]
Just tested this out on `rosenberg`.

* Pulled your branch
* conda activated
* updated the env using `conda env update --file environment.yml --prune`
* `pip install -e .` updated
* In another tab, started monitoring the nvidia usage `watch -n 1 -d nvidia-smi`
* In the original tab, tested out three scenarios
  * axondeepseg -t SEM -i AxonDeepSeg/models/model_seg_rat_axon-myelin_sem/data_test/image.png --gpu-id 5
    * Only GPU 5 was used throughout the run
  * CUDA_VISIBLE_DEVICES=5 axondeepseg -t SEM -i AxonDeepSeg/models/model_seg_rat_axon-myelin_sem/data_test/image.png --gpu-id 0
    * Only GPU 5 was used throughout the run
  * CUDA_VISIBLE_DEVICES=5,6 axondeepseg -t SEM -i AxonDeepSeg/models/model_seg_rat_axon-myelin_sem/data_test/image.png --gpu-id 1
    * Only GPU 6 was used throughout the run

So it looks like all issues have been resolved, and the feature works as expected through the CLI! 

@Stoyan-I-A will you be testing the FSLeyes plugin? @mariehbourget have you tried it after your update?

**Comment by mariehbourget** [comment]
> @mariehbourget have you tried it after your update?

I only tried the CLI before.
I just tried fsleyes on rosenberg with x-forwarding (`ssh -X username@rosenberg.neuro.polymtl.ca`) and it works as expected as well.
i.e. the number of GPU available in the settings menu is consistent with CUDA_VISIBLE_DEVICES and only the selected GPU is used throughout the run.

**Comment by Stoyan-I-A** [comment]
I haven't tested it, but the code looks good to me.

**Pull request merged**
